WordPress/openverse-frontend
# Add recent searches to the search store

## Fixes


Related to #479 by @zackkrida 

## Description



This PR adds a user's recent searches to the search store, and syncs these searches to localstorage. Adding these searches to a global store _and_ localstorage is the best way to ensure the recent searches can be stored and read from different components without doing a lot of convoluted work to sync state.

This code includes some work to make sure duplicate searches aren't stored in the history and to cap the number of saved searches to the env-variable set default.

It was interesting to learn about how Pinia unwraps reactives with more detail while working on this. Really wishing we were on Nuxt 3 and Vue 3 with the hopes it would fix some odd behavior I had to deal with here! Being stuck on the really old (`0.2.1`) `@pinia/nuxt` package feels pretty brittle.

## Testing Instructions


Make some searches in Openverse and console.log `localStorage.getItem('recent-searches')` after to ensure they saved correctly. Make sure they never exceed the currently set limit of 4 searches and that you don't see any duplicates.

## Todo

The steps remaining to undraft this are:

- [x] Put the saving of searches behind a feature flag (it seems a bit sneaky to store these for users until the UI is live)
- [x] Update search-store tests for these new actions
- [x] Sort out bug where searches don't save after homepage navigation

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
